### PR TITLE
Update: Let `no-restricted-properties` check destructuring (fixes #7147)

### LINKED ISSUE #7147: `no-restricted-properties`: does not work with destructuring
**Tell us about your environment**

* **ESLint Version:**
v3.5.0
* **Node Version:**
v6.5.0
* **npm Version:**
3.something

**What parser (default, Babel-ESLint, etc.) are you using?**
default

**Please show your full configuration:**
n/a

**What did you do? Please include the actual source code causing the issue.**
```sh
echo 'const { bar } = foo; const bar = foo.bar;' | ./node_modules/.bin/eslint --stdin --parser-options '{"ecmaVersion": 6}' --rule 'no-restricted-properties:[2,{"object":"foo","property":"bar"}]' --rule no-undef:0 --rule no-unused-vars:0 --no-eslintrc
```

**What did you expect to happen?**
```sh
<text>
  1:##  error  'foo.bar' is restricted from being used  no-restricted-properties
  1:34  error  'foo.bar' is restricted from being used  no-restricted-properties

✖ 2 problems (2 errors, 0 warnings)
```

**What actually happened? Please include the actual, raw output from ESLint.**
```sh
<text>
  1:34  error  'foo.bar' is restricted from being used  no-restricted-properties

✖ 1 problem (1 error, 0 warnings)
```

Per https://github.com/eslint/eslint/pull/7139#issuecomment-246749114

### PR DESCRIPTION
**Pull request by not-an-aardvark**
**What is the purpose of this pull request? (put an "X" next to item)**

```
[ ] Documentation update
[x] Bug fix ([template](https://github.com/eslint/eslint/blob/master/templates/bug-report.md))
[ ] New rule ([template](https://github.com/eslint/eslint/blob/master/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://github.com/eslint/eslint/blob/master/templates/rule-change-proposal.md))
[ ] Add autofixing to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:
```

See #7147 for the template answers.

**Please check each item to ensure your pull request is ready:**

- [x] I've read the [pull request guide](http://eslint.org/docs/developer-guide/contributing/pull-requests)
- [x] I've included tests for my change
- [x] I've updated documentation for my change (if appropriate)

**What changes did you make? (Give an overview)**

This allows `no-restricted-properties` to report errors from destructuring patterns.

**Is there anything you'd like reviewers to focus on?**

Nothing in particular.

~~This PR is built on top of the commit from #7139, since it includes some test cases that correspond to the options in #7139. It shouldn't be merged until after #7139 is merged.~~ (edit: #7139 is merged now.)

~~In the meantime, see [here](https://github.com/eslint/eslint/pull/7151/commits/311e93384f4b999d472e5e75a74e59f4856b97c0) for the diff that only includes changes from this PR.~~

### CODE REVIEW

File: lib/rules/no-restricted-properties.js [hunk 1]
```diff
@@ -51,36 +68,79 @@ module.exports = {
...
                 const objectName = node.object && node.object.name;
                 const propertyName = astUtils.getStaticPropertyName(node);
-                const matchedObject = restrictedProperties.get(objectName);
-                const matchedObjectProperty = matchedObject && matchedObject.get(propertyName);
 
-                if (matchedObjectProperty) {
-                    const message = matchedObjectProperty.message ? " " + matchedObjectProperty.message : "";
+                checkPropertyAccess(node, objectName, propertyName);
+            },
+            VariableDeclarator(node) {
```
Comment (mysticatea) [inline lib/rules/no-restricted-properties.js]: And `AssignmentExpression`/`AssignmentPattern` can access properties by destructuring.

Comment (not-an-aardvark) [inline lib/rules/no-restricted-properties.js]: I updated the PR to check for an `AssignmentExpression`, but what is an example of an `AssignmentPattern`?

Comment (mysticatea) [inline lib/rules/no-restricted-properties.js]:
```js
const obj = {a: 1}
function foo({a} = obj) { // obj.a
}
```

Comment (not-an-aardvark) [inline lib/rules/no-restricted-properties.js]: Updated to check for `AssignmentPattern`s as well.

**Comment by not-an-aardvark** [comment]
On second thought, I realized maybe we shouldn't consider `var [bar] = foo` to be accessing the `0` property of `foo` since that's only true for arrays. It would be a false positive for any other iterable:

```js
/* eslint no-restricted-properties: [2, {"object": "foo", "property": "0"}] */

var foo = new Set();

var [bar] = foo; // reports an error with this PR, but it probably shouldn't.
```

So we should only consider `ObjectPattern` destructuring and not `ArrayPattern`.

File: lib/rules/no-restricted-properties.js (hunk 1 above)
Comment (nzakas) [inline lib/rules/no-restricted-properties.js]: @mysticatea does this rule flag this?

```js
function foo(obj) {
    return obj.a;
}
```

Comment (not-an-aardvark) [inline lib/rules/no-restricted-properties.js]: @nzakas: Yes; it reports any reference to the `a` property of anything named `obj` (assuming the config is `[2, {"object": "obj", "property": "a"}]` ).

Comment (nzakas) [inline lib/rules/no-restricted-properties.js]: Cool, just wanted to double-check. :+1:

**Comment by platinumazure** [comment]
@not-an-aardvark We can consider supporting array destructure later-- I think supporting object destructure is the most important thing here.

File: lib/rules/no-restricted-properties.js (hunk 1 above)
Comment (mysticatea) [inline lib/rules/no-restricted-properties.js]: Yes. It's checked as a `MemberExpression`.

File: lib/rules/no-restricted-properties.js [hunk 2]
```diff
@@ -51,38 +68,96 @@ module.exports = {
...
+                    const objectName = node.init.name;
 
-                    context.report(node, "'{{objectName}}.{{propertyName}}' is restricted from being used.{{message}}", {
-                        objectName,
-                        propertyName,
-                        message
-                    });
+                    if (node.id.type === "ObjectPattern") {
+                        node.id.properties.forEach(property => {
+                            checkPropertyAccess(node.id, objectName, property.key.name || property.key.value);
```
Comment (mysticatea) [inline lib/rules/no-restricted-properties.js]: This should ignore the property if the property has dynamic name.

```js
const name = astUtils.getStaticPropertyName(property);
if (name) {
    checkPropertyAccess(node.id, objectName, name);
}
```

There is the same code in `checkDestructuringAssignment` function.

File: lib/rules/no-restricted-properties.js [hunk 3]
```diff
@@ -22,22 +22,39 @@ module.exports = {
...
+                            },
+                            property: {
+                                type: "string"
+                            },
+                            message: {
+                                type: "string"
+                            }
+                        },
+                        additionalProperties: false,
+                        required: ["object"]
```
Comment (mysticatea) [inline lib/rules/no-restricted-properties.js]: By the way, I'm wondering if `property` should be omittable because this rule is `no-restricted-properties`

Comment (mysticatea) [inline lib/rules/no-restricted-properties.js]: Ah, sorry, never mind...

**Comment by mysticatea** [comment]
LGTM, thank you!
Just waiting for #7139 is accepted.

**Pull request merged**